src/www/app/app.ts

```diff
@@ -153,83 +153,91 @@ export class App {
+    } else if (error instanceof errors.SystemConfigurationException) {
+      toastMessage = this.localize('outline-plugin-error-system-configuration');
+    } else if (error instanceof errors.ShadowsocksUnsupportedCipher) {
+      toastMessage = this.localize('error-shadowsocks-unsupported-cipher', 'cipher', error.cipher);
+    } else if (error instanceof errors.ServerAccessKeyInvalid) {
+      toastMessage = this.localize('error-connection-configuration');
+      buttonMessage = this.localize('error-details');
+      buttonHandler = () => {
+        this.showErrorDetailDialog(error);
+      };
+    } else if (error instanceof errors.SessionConfigFetchFailed) {
+      toastMessage = this.localize('error-connection-configuration-fetch');
+      buttonMessage = this.localize('error-details');
+      buttonHandler = () => {
+        this.showErrorDetailDialog(error);
+      };
+    } else if (error instanceof errors.ProxyConnectionFailure) {
+      toastMessage = this.localize('error-connection-proxy');
+      buttonMessage = this.localize('error-details');
+      buttonHandler = () => {
+        this.showErrorDetailDialog(error);
+      };
     } else {
-      const hasErrorDetails = Boolean(e.message || e.cause);
-
-      messageKey = 'error-unexpected';
-      buttonKey = hasErrorDetails ? 'error-details' : undefined;
-      buttonHandler = hasErrorDetails
-        ? () => {
-            this.showErrorDetailDialog(e);
-          }
-        : undefined;
+      const hasErrorDetails = Boolean(error.message || error.cause);
+      toastMessage = this.localize('error-unexpected');
+
+      if (hasErrorDetails) {
+        buttonMessage = this.localize('error-details');
+        buttonHandler = () => {
+          this.showErrorDetailDialog(error);
+        };
+      }
     }
 
     // Defer by 500ms so that this toast is shown after any toasts that get shown when any
     // currently-in-flight domain events land (e.g. fake servers added).
     if (this.rootEl && this.rootEl.async) {
       this.rootEl.async(() => {
-        const buttonMessage = typeof buttonKey === 'string' ? this.localize(buttonKey) : undefined;
-        const hasButtonMessage = Boolean(buttonMessage);
-
-        if (hasButtonMessage) {
-          this.rootEl.showToast(
-            this.localize(messageKey, ...messageParams),
-            toastDuration,
-            buttonMessage,
-            buttonHandler,
-            buttonLink
-          );
-        } else {
-          this.rootEl.showToast(this.localize(messageKey, ...messageParams), toastDuration);
-        }
+        toastMessage = toastMessage ?? error.message;
+        const toastArguments =
+          typeof buttonMessage === 'string'
+            ? [toastMessage, toastDuration, buttonMessage, buttonHandler, buttonLink]
+            : [toastMessage, toastDuration];
+
+        this.rootEl.showToast(...toastArguments);
```

Comment:
This is more readable:
```suggestion
       if (typeof buttonMessage === 'string') {
         this.rootEl.showToast(toastMessage, toastDuration, buttonMessage, buttonHandler, buttonLink);
       } else {
         this.rootEl.showToast(toastMessage, toastDuration);
       }
```